client/src/main/scala/org/http4s/client/middleware/FollowRedirect.scala

```diff
@@ -94,23 +94,25 @@ object FollowRedirect {
     }
 
     def prepareLoop(req: Request[F], redirects: Int): F[Resource[F, Response[F]]] =
-      F.continual(client.run(req).allocated) {
-        case Right((resp, dispose)) =>
-          (methodForRedirect(req, resp), resp.headers.get(Location)) match {
-            case (Some(method), Some(loc)) if redirects < maxRedirects =>
-              val nextReq = nextRequest(req, loc.uri, method, resp.cookies)
-              dispose >> prepareLoop(nextReq, redirects + 1).map(_.map { response =>
-                // prepend because `prepareLoop` is recursive
-                response.withAttribute(redirectUrisKey, nextReq.uri +: getRedirectUris(response))
-              })
-            case _ =>
-              // IF the response is missing the Location header, OR there is no method to redirect,
-              // OR we have exceeded max number of redirections, THEN we redirect no more
-              Resource.make(resp.pure[F])(_ => dispose).pure[F]
-          }
-
-        case Left(e) =>
-          F.raiseError(e)
+      F.uncancelable { _ =>
+        client.run(req).allocated.attempt.flatMap {
```

Comment:
I never liked this `.allocated`, but I think it's still necessary.  We need to dispose of the connection before allocating another, or else we can eat up all the connections.

I wonder whether we could extract some sort of "resource loop" combinator to cats-effect.  This was a hard thing to get right.

Reply:
I'll take a closer look and see if there's an obvious refactoring now that the majority of the work is done :)

Reply:
Check it out now, it uses the `Hotswap` type which does exactly what you said -- linearly allocate a sequence of resources, ensuring each one is deallocated before the next is allocated.

Reply:
Oh, yes, that's very nice!  There's a similar loop in `Retry` that could get the same treatment, I think.